tests/crypto_test.py

```diff
@@ -95,3 +153,212 @@ def test_verify_signature_bad(self):
+
+    def test_encrypt(self):
+        with self.assertRaises(GPGProblem) as caught:
+            crypto.validate_key(make_key(can_encrypt=False), encrypt=True)
+
+        self.assertEqual(caught.exception.code, GPGCode.KEY_CANNOT_ENCRYPT)
+
+    def test_encrypt_no_check(self):
+        try:
+            crypto.validate_key(make_key(can_encrypt=False))
+        except GPGProblem as e:
+            raise AssertionError(e)
+
+    def test_sign(self):
+        with self.assertRaises(GPGProblem) as caught:
+            crypto.validate_key(make_key(can_sign=False), sign=True)
+
+        self.assertEqual(caught.exception.code, GPGCode.KEY_CANNOT_SIGN)
+
+    def test_sign_no_check(self):
+        try:
+            crypto.validate_key(make_key(can_sign=False))
+        except GPGProblem as e:
+            raise AssertionError(e)
+
+
+class TestCheckUIDValidity(unittest.TestCase):
+
+    def test_valid_single(self):
+        key = make_key()
+        key.uids[0] = make_uid(mock.sentinel.EMAIL)
+        ret = crypto.check_uid_validity(key, mock.sentinel.EMAIL)
+        self.assertTrue(ret)
+
+    def test_valid_multiple(self):
+        key = make_key()
+        key.uids = [
+            make_uid(mock.sentinel.EMAIL),
+            make_uid(mock.sentinel.EMAIL1),
+        ]
+
+        ret = crypto.check_uid_validity(key, mock.sentinel.EMAIL1)
+        self.assertTrue(ret)
+
+    def test_invalid_email(self):
+        key = make_key()
+        key.uids[0] = make_uid(mock.sentinel.EMAIL)
+        ret = crypto.check_uid_validity(key, mock.sentinel.EMAIL1)
+        self.assertFalse(ret)
+
+    def test_invalid_revoked(self):
+        key = make_key()
+        key.uids[0] = make_uid(mock.sentinel.EMAIL, revoked=True)
+        ret = crypto.check_uid_validity(key, mock.sentinel.EMAIL)
+        self.assertFalse(ret)
+
+    def test_invalid_invalid(self):
+        key = make_key()
+        key.uids[0] = make_uid(mock.sentinel.EMAIL, invalid=True)
+        ret = crypto.check_uid_validity(key, mock.sentinel.EMAIL)
+        self.assertFalse(ret)
+
+    def test_invalid_not_enough_trust(self):
+        key = make_key()
+        key.uids[0] = make_uid(mock.sentinel.EMAIL,
+                               validity=gpgme.VALIDITY_UNDEFINED)
+        ret = crypto.check_uid_validity(key, mock.sentinel.EMAIL)
+        self.assertFalse(ret)
+
+
+class TestListKeys(WithKey):
+
+    def test_list_no_hints(self):
+        # This only tests that you get 3 keys back (the number in our test
+        # keyring), it might be worth adding tests to check more about the keys
+        # returned
+        values = crypto.list_keys()
+        self.assertEqual(len(list(values)), 3)
+
+    def test_list_hint(self):
+        values = crypto.list_keys(hint="ambig")
+        self.assertEqual(len(list(values)), 2)
+
+
+class TestGetKey(WithKey):
+
+    def test_plain(self):
+        # Test the uid of the only identity attached to the key we generated.
+        ctx = gpgme.Context()
```

Comment:
They're each new ones. Originally I'd attached it to the class level fixture to save some typing, but there isn't a need to. A context isn't a new keyring, it's more of a view into gpg, and theoretically allows you to do some async stuff since Context is implemented in C and doesn't take the gil. I'm not sure that it really matters one way or another, but I've done away with the `WithKey` class, since it was basically unused and moved the fingerprint of the key to a module level variable.